crates/bevy_render/src/render_component.rs

```diff
@@ -11,6 +11,7 @@ use bevy_ecs::{
     query::{QueryItem, WorldQuery},
     system::{lifetimeless::Read, StaticSystemParam},
 };
+use encase::{internal::WriteInto, ShaderType};
```

Comment:
Part of the rationale behind vendoring crevice was so we could re-export derives (by hard-coding bevy paths). ShaderType is common enough for shader workflows that users should be able to derive it without manually pulling in encase. This is a pretty hard requirement from my perspective. Bytemuck would be nice, but its a _slightly_ lower priority imo.

Additional user-facing derive complexity isn't an option (which many of the links above introduce). So from my perspective our options are:
1. Fully vendor encase
2. Only vendor encase derives / export our own "mirror" of the macro with our own paths. This would make it easy for things to get "out of sync" so we'd need to be _very_ careful (and vigilant). One wrong patch release on the encase side could completely break the bevy derive. Us forgetting to update the derive when we update our encase import could also break things (in potentially hard to identify ways).
3. encase adds a "bevy-macro-path" feature that provides opt-in upstream support for bevy paths. This adds upstream complexity, but would enable us to use encase upstream directly without full or partial vendoring. Thats a call solely in the hands of @teoxoy and I won't apply any more pressure here. 

There might be a fourth option where encase's derive crate exports a parameterized derive helper function, so all we need to do is write a simple "wrapper derive" over the top of it that passes the bevy paths in as arguments. This would ensure we're always in sync with whatever version of encase we're using. I'm not aware of any precedent for that, so I'm not actually sure if this would work. Might be worth testing though! Maybe we could pioneer a new Rust pattern!

Reply:
That fourth option sounds good... we need it ourselves already for some cases where a macro is used internally referring to internal crates as `bevy_crate` vs externally `bevy::crate`. Maybe I found a hack around that already, I don't remember, but I do remember hitting that as a problem. Probably when trying to use a macro within the crate where it is defined.

I don't like 1 nor 2. 4 sounds more maintainable than 3, but either would work. And yeah, totally up to @teoxoy but I figure it's good to express preferences/thoughts in case you're on the fence and find it helpful. :)

Reply:
@cart @superdump thanks for the insights! I was under the assumption that crevice was vendored because you guys improved some things and fixed some bugs that didn't get integrated into it in a timely manner.

As I previously mentioned in the issue, the reason I was slightly opposed to 3 was because feature flags are supposed to be compatible with each other (i.e. not change behavior in a breaking change way). The scenario would be: encase **without** the bevy feature flag is part of a project's dependencies (directly or indirectly) together with bevy (and indirectly encase **with** the feature flag) it would cause the project to not compile since when cargo does the feature unification we will end up only with the version of encase **with** the bevy feature flag causing uses of the `ShaderType` macro that's directly imported from encase to no longer work.

All that being said, I think a different way of doing 3 might work. If encase adds a 2nd derive macro `BevyShaderType` that would get re-exported as `ShaderType` in bevy (this would be a cut down version of 4 I guess).

Regarding 4, (if I understood correctly) I don't think that's possible since a `proc-macro` crate can't export any items other than functions tagged with `#[proc_macro]`, `#[proc_macro_derive]`, or `#[proc_macro_attribute]` (that's the actual error message).

Unless I misunderstood how 4 should work, I'd say we go with the variation of 3 I proposed above (I think that one should work just fine).

Reply:
I managed to get 4 working! I'm quite happy with the way it turned out. I think it's the most elegant solution to this problem.

See
https://github.com/bevyengine/bevy/pull/4339/commits/ef7541361644972bc57112e0aa7a612b47031382
https://github.com/teoxoy/encase/commit/a2c202d4405430e8b9f21e7b1104a13d6c0779c9